File: server-rs/src/settings/network_config.rs
use crate::db::get_data_dir;
use crate::error::{AppError, AppResult};
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::sync::LazyLock;

pub const DEFAULT_NETWORK_HOST: &str = "workbench.local";
pub const DEFAULT_NETWORK_PORT: u16 = 4738;

static HOST_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$")
        .expect("valid host regex")
});

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct NetworkConfig {
    pub host: String,
    pub port: u16,
}

pub type PatchNetworkConfig = NetworkConfigPatch;

#[derive(Debug, Clone, Default, Deserialize)]
pub struct NetworkConfigPatch {
    pub host: Option<String>,
    pub port: Option<u16>,
}

#[derive(Debug, Clone, Default)]
pub struct ResolveNetworkOptions {
    pub port: Option<u16>,
    pub host: Option<String>,
}

pub fn get_network_config_path() -> PathBuf {
    get_data_dir().join("config.json")
}

pub fn default_network_config() -> NetworkConfig {
    NetworkConfig {
        host: DEFAULT_NETWORK_HOST.to_string(),
        port: DEFAULT_NETWORK_PORT,
    }
}

fn validate_network_config(config: &NetworkConfig) -> AppResult<()> {
    if config.host.is_empty() || config.host.len() > 253 {
        return Err(AppError::BadRequest("Invalid hostname".into()));
    }
    if !HOST_PATTERN.is_match(&config.host) {
        return Err(AppError::BadRequest("Invalid hostname".into()));
    }
    if config.port == 0 {
        return Err(AppError::BadRequest("Invalid port".into()));
    }
    Ok(())
}

pub fn load_network_config() -> NetworkConfig {
    let path = get_network_config_path();
    if !path.exists() {
        return default_network_config();
    }

    let raw = match fs::read_to_string(&path) {
        Ok(raw) => raw,
        Err(_) => return default_network_config(),
    };

    match serde_json::from_str::<NetworkConfig>(&raw) {
        Ok(config) if validate_network_config(&config).is_ok() => config,
        _ => default_network_config(),
    }
}

pub fn save_network_config(patch: &NetworkConfigPatch) -> AppResult<NetworkConfig> {
    let current = load_network_config();
    let next = NetworkConfig {
        host: patch.host.clone().unwrap_or(current.host),
        port: patch.port.unwrap_or(current.port),
    };
    validate_network_config(&next)?;

    let dir = get_data_dir();
    fs::create_dir_all(&dir).map_err(|e| AppError::Internal(e.to_string()))?;
    let json = serde_json::to_string_pretty(&next)
        .map_err(|e| AppError::Internal(e.to_string()))?;
    fs::write(get_network_config_path(), format!("{json}\n"))
        .map_err(|e| AppError::Internal(e.to_string()))?;
    Ok(next)
}

pub fn resolve_network_config(options: &ResolveNetworkOptions) -> AppResult<NetworkConfig> {
    let file = load_network_config();
    let env_port = std::env::var("PORT")
        .ok()
        .and_then(|p| p.parse::<u16>().ok());
    let env_host = std::env::var("WORKBENCH_HOST")
        .ok()
        .map(|h| h.trim().to_string())
        .filter(|h| !h.is_empty());

    let port = options.port.or(env_port).unwrap_or(file.port);
    let host = options
        .host
        .clone()
        .or(env_host)
        .unwrap_or_else(|| file.host.clone());

    let config = NetworkConfig { host, port };
    validate_network_config(&config)?;
    Ok(config)
}

pub fn hosts_file_line(host: &str) -> String {
    format!("127.0.0.1 {host}")
}

pub fn config_differs_from_running(saved: &NetworkConfig, running: &NetworkConfig) -> bool {
    saved.host != running.host || saved.port != running.port
}

File: server/modules/settings/network-config.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { z } from "zod";
import { getDataDir } from "../../db/data-dir.js";

export const DEFAULT_NETWORK_HOST = "workbench.local";
export const DEFAULT_NETWORK_PORT = 4738;
/** Added to prod port for feature-branch / non-default worktree instances. */
export const NON_PROD_PORT_OFFSET = 1;

export function nonProdPort(prodPort: number): number {
  return prodPort + NON_PROD_PORT_OFFSET;
}

const HOST_PATTERN =
  /^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$/i;

const networkConfigSchema = z.object({
  host: z
    .string()
    .min(1)
    .max(253)
    .refine((h) => HOST_PATTERN.test(h), "Invalid hostname"),
  port: z.number().int().min(1).max(65535),
});

export type NetworkConfig = z.infer<typeof networkConfigSchema>;

export type PatchNetworkConfig = Partial<NetworkConfig>;

export function getNetworkConfigPath(): string {
  return join(getDataDir(), "config.json");
}

export function defaultNetworkConfig(): NetworkConfig {
  return { host: DEFAULT_NETWORK_HOST, port: DEFAULT_NETWORK_PORT };
}

export function loadNetworkConfig(): NetworkConfig {
  const path = getNetworkConfigPath();
  if (!existsSync(path)) return defaultNetworkConfig();
  try {
    const raw = JSON.parse(readFileSync(path, "utf8")) as unknown;
    const parsed = networkConfigSchema.safeParse(raw);
    if (parsed.success) return parsed.data;
  } catch {
    // fall through
  }
  return defaultNetworkConfig();
}

export function saveNetworkConfig(patch: PatchNetworkConfig): NetworkConfig {
  const current = loadNetworkConfig();
  const next = networkConfigSchema.parse({ ...current, ...patch });
  const dir = getDataDir();
  mkdirSync(dir, { recursive: true });
  writeFileSync(getNetworkConfigPath(), `${JSON.stringify(next, null, 2)}\n`, "utf8");
  return next;
}

export interface ResolveNetworkOptions {
  port?: number;
  host?: string;
}

/** Merge saved config with env and explicit overrides (CLI). */
export function resolveNetworkConfig(options: ResolveNetworkOptions = {}): NetworkConfig {
  const file = loadNetworkConfig();
  const envPort = process.env.PORT ? parseInt(process.env.PORT, 10) : undefined;
  const envHost = process.env.WORKBENCH_HOST?.trim() || undefined;

  const port = options.port ?? (Number.isFinite(envPort) ? envPort! : file.port);
  const host = options.host ?? envHost ?? file.host;

  return networkConfigSchema.parse({ host, port });
}

export function hostsFileLine(host: string): string {
  return `127.0.0.1 ${host}`;
}

export function configDiffersFromRunning(
  saved: NetworkConfig,
  running: NetworkConfig,
): boolean {
  return saved.host !== running.host || saved.port !== running.port;
}
